geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/KryoFeatureSerialization.scala

```diff
@@ -47,155 +49,191 @@ trait KryoFeatureSerialization extends SimpleFeatureSerializer {
     }
   }
 
-  private def writeNullable(wrapped: (Output, AnyRef) => Unit): (Output, AnyRef) => Unit = {
-    (o: Output, v: AnyRef) => {
-      if (v == null) {
-        o.write(NULL_BYTE)
-      } else {
-        o.write(NON_NULL_BYTE)
-        wrapped(o, v)
+  sealed trait KryoAttributeWriter {
+    def apply(output: Output, value: AnyRef): Unit
+  }
+
+  case object KryoStringWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeString(value.asInstanceOf[String])
+  }
+
+  case object KryoIntWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeInt(value.asInstanceOf[Int])
+  }
+
+  case object KryoLongWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeLong(value.asInstanceOf[Long])
+  }
+
+  case object KryoFloatWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeFloat(value.asInstanceOf[Float])
+  }
+
+  case object KryoDoubleWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeDouble(value.asInstanceOf[Double])
+  }
+
+  case object KryoBooleanWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeBoolean(value.asInstanceOf[Boolean])
+  }
+
+  case object KryoDateWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = output.writeLong(value.asInstanceOf[Date].getTime)
+  }
+
+  case object KryoUuidWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = {
+      val uuid = value.asInstanceOf[UUID]
+      output.writeLong(uuid.getMostSignificantBits)
+      output.writeLong(uuid.getLeastSignificantBits)
+    }
+  }
+
+  case object KryoJsonWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit =
+      KryoJsonSerialization.serialize(output, value.asInstanceOf[String])
+  }
+
+  case object KryoBytesWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit = {
+      val array = value.asInstanceOf[Array[Byte]]
+      output.writeInt(array.length, true)
+      output.writeBytes(array)
+    }
+  }
+
+  case object KryoGeometryWkbWriter extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit =
+      KryoGeometrySerialization.serializeWkb(output, value.asInstanceOf[Geometry])
+  }
+
+  case class KryoGeometryTwkbWriter(precision: GeometryPrecision.TwkbPrecision) extends KryoAttributeWriter {
+    override def apply(output: Output, value: AnyRef): Unit =
+      KryoGeometrySerialization.serialize(output, value.asInstanceOf[Geometry], precision)
+  }
+
+  case class KryoListWriter(elements: KryoAttributeWriter) extends KryoAttributeWriter {
+    // TODO handle null elements?
```

Comment:
Do we have unit tests which cover this?

Reply:
no, I don't think that we had any tests for this, and I didn't add any. do you think we should support null values in a list or map? I can add it, but it would be some extra bytes per list element.